<a href="https://colab.research.google.com/github/Perlangas28/Simulacion_1/blob/main/LineaDeEsperaConUnServidor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Tarea**: Línea de espera con un servidor


**Alumna: Perla Blancas Hernández**  
**Materia:** Simulación 1
**Prueba aplicada:** Sheldon Ross  

Programar en Jupyter Notebook el seudocódigo propuesto por Sheldon Ross.
Como prueba, usar $λ=4$ y $μ=6$. Comentar y documentar.
Comparar las soluciones analítica y por simulación

In [ ]:
import numpy as np
import pandas as pd

Primero definimos los parámetros:

$λ=4:$ Llegan en promedio 4 clientes por unidad de tiempo.

$μ=6:$ El servidor puede atender en promedio 6 clientes por unidad de tiempo.

$N=100000:$ Vamos a simular 100000 clientes.

In [ ]:
lam = 4.0      # tasa de llegada
mu = 6.0       # tasa de servicio
N = 100000     # número de clientes

np.random.seed(123)

Creamos los arreglos para que se almacenen los tiempos de llegada

In [ ]:
arrival_times = np.zeros(N)
service_times = np.random.exponential(1/mu, N)

Ahora generamos los tiempos de llegada.

In [ ]:
interarrivals = np.random.exponential(1/lam, N)

Para calcular los tiempos de llegada utilizaremos:

$A_i = A_{i-1} + T_i$

Donde:

$A_i$ : Tiempo de llegada

$T_i$ : Tiempo entre llegadas

In [ ]:
for i in range(1, N):
    arrival_times[i] = arrival_times[i-1] + interarrivals[i]

Ahora hacemos lo mismo pero para los tiempos de salida.

Para el inicio de servicio, el cliente puede comenzar a atenderse cuando ocurren ambas cosas:

1. Ya llegó
2. El servidor quedó libre, entonces usamos:

$B_i = max(A_i, D_{i-1})$

donde:

$A_i$: Llegada

$D_{i-1}$: Salida del ciente anterior

$B_i$: Inicio del servicio

Para el tiempo de salida:

$D_i = B_i + S_i$

Donde:

$D_i$: Salida

$B_i$: Inicio del servicio

$S_i$: Tiempo de servicio


In [ ]:
# Variables de simulación
start_service = np.zeros(N)
departure_times = np.zeros(N)

# Primer cliente
start_service[0] = arrival_times[0]
departure_times[0] = start_service[0] + service_times[0]

# Algoritmo de Ross
for i in range(1, N):

    # inicio de servicio
    start_service[i] = max(arrival_times[i],
                           departure_times[i-1])

    # salida
    departure_times[i] = start_service[i] + service_times[i]

Después, calculamos el tiempo de espera en cola, el tiempo total en el sistema y los promedios.

El tiempo de espera lo calculamos de la siguiente manera: (Inicio - llegada)

$W_q = B_i - A_i$

Para el tiempo total en el sistema: (Salida - Llegada)

$W = D_i - A_i$

Finalmente calculamos los promedios simulados:

$ Wq = ∑W_{qi}/N$

y

$W = ∑W_i/N $


In [ ]:
waiting_queue = start_service - arrival_times

time_system = departure_times - arrival_times

Wq_sim = np.mean(waiting_queue)
W_sim = np.mean(time_system)

Lq_sim = lam * Wq_sim
L_sim = lam * W_sim

Resultados finales:

Aquí usamos las fórmulas analíticas siguientes:

$L = ρ / 1 - ρ$

$L_q = ρ^2 / 1 - ρ$

$W = 1 / μ - λ$

$W_q = ρ / μ - λ$

In [ ]:
resultados = pd.DataFrame({
    "Medida":["L","Lq","W","Wq"],
    "Analítico":[
        (lam/mu)/(1-lam/mu),
        (lam/mu)**2/(1-lam/mu),
        1/(mu-lam),
        (lam/mu)/(mu-lam)
    ],
    "Simulación":[
        L_sim,
        Lq_sim,
        W_sim,
        Wq_sim
    ]
})

print(resultados)

  Medida  Analítico  Simulación
0      L   2.000000    1.983833
1     Lq   1.333333    1.317045
2      W   0.500000    0.495958
3     Wq   0.333333    0.329261


Los resultados obtenidos mediante simulación son muy cercanos a los valores analíticos.
Para $λ=4$ y $μ=6$, el sistema mantiene en promedio 2 clientes en el sistema, 1.33 clientes en la cola, un tiempo promedio en el sistema de 0.5 unidades de tiempo y una espera promedio en cola de 0.333 unidades de tiempo, coincidiendo estrechamente con los valores obtenidos mediante simulación.